### PR TITLE
HPCC-24712 Ignore a TLK index root that has a single row in it

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Smoketest

### CODE REVIEW

**Comment by richardkchapman** [comment]
@richardkchapman

File: system/jhtree/jhtree.cpp [hunk 1]
```diff
@@ -992,6 +992,12 @@ void CKeyIndex::init(KeyHdr &hdr, bool isTLK, bool allowPreload)
             }
         }
         rootNode = nodeCache->getNode(this, iD, rootPos, NULL, isTLK);
+        if (rootNode && isTLK && !rootNode->isLeaf() && rootNode->getNumKeys()==1)   // In theory should work for non-TLK keys too, but not been tested and less likely to happen
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp:995]: Could also do with a comment similar to the commit text.

**Comment by ghalliday** [comment]
@richardkchapman

File: system/jhtree/jhtree.cpp [hunk 2]
```diff
@@ -992,6 +992,12 @@ void CKeyIndex::init(KeyHdr &hdr, bool isTLK, bool allowPreload)
...
+        {
+            Linked<CJHTreeNode> oldRoot = rootNode;
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: I think you are right - changed

File: system/jhtree/jhtree.cpp (hunk 1 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: added

**Comment by richardkchapman** [comment]
@ghalliday

File: system/jhtree/jhtree.cpp [hunk 3]
```diff
@@ -992,6 +992,18 @@ void CKeyIndex::init(KeyHdr &hdr, bool isTLK, bool allowPreload)
             }
         }
         rootNode = nodeCache->getNode(this, iD, rootPos, NULL, isTLK);
+		// It's not uncommon for a TLK to have a "root node" that has a single entry in it pointing to a leaf node
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp:995]: tabs inserted rather than spaces.

**Pull request merged**